ArgoProject/src/systems/CollisionSystem.cpp

```diff
@@ -0,0 +1,81 @@
+#include "stdafx.h"
+#include "CollisionSystem.h"
+
+// components
+#include "components/Location.h"
+#include "components/Dimensions.h"
+#include "components/Render.h"
+#include "components/Collision.h"
+#include "components/Input.h"
+#include "components/Motion.h"
+
+app::sys::CollisionSystem::CollisionSystem()
+	: BaseSystem()
+{
+	//prepare these components
+	m_registry.prepare<comp::Collision, comp::Location, comp::Input, comp::Dimensions>();
+}
+
+void app::sys::CollisionSystem::update(app::time::seconds const & dt)
+{
+	//type of collision
+	std::string manifoldType;
+
+	//view player
+	m_registry.view<comp::Collision, comp::Input, comp::Location, comp::Dimensions>(entt::persistent_t())
+		.each([&, this](app::Entity const entity, comp::Collision & collision, comp::Input & input, comp::Location & location, comp::Dimensions & dimensions)
+	{
+		//update collision boxes
+		updateAABB(std::get<cute::c2AABB>(collision.collisionBox), location.position, dimensions.size, dimensions.origin);
+		//view everything with collisions
+		m_registry.view<comp::Collision, comp::Location, comp::Dimensions>(entt::persistent_t())
+			.each([&, this](app::Entity const secEntity, comp::Collision & secCollision, comp::Location & secLocation, comp::Dimensions & secDimensions)
+		{
+			//if we are not the player
+			if (entity != secEntity)
+			{
+				updateAABB(std::get<cute::c2AABB>(secCollision.collisionBox), secLocation.position, secDimensions.size, secDimensions.origin);
+				
+				manifoldType = checkManifoldType(collision.collisionBox, secCollision.collisionBox);
```

Comment:
std::variant best practice use `std::visit` function in which you pass in a object that has a overloaded member function `operator()` that accepts all the types that the variant contains.
```suggestion
	cute::c2Manifold app::vis::CollisionBoundsManifoldVisitor::collisionBetween(std::variant<cute::c2AABB, cute::c2Circle> const & left, std::variant<cute::c2AABB, cute::c2Circle> const & right)
	{
		auto manifold = cute::c2Manifold();
		std::visit(vis::CollisionBoundsManifoldVisitor(manifold, right), left);
		return manifold;
	}
```

```cpp
class CollisionBoundsManifoldVisitor
{
	CollisionBoundsManifoldVisitor(cute::c2Manifold & manifold, std::variant<cute::c2AABB, cute::c2Circle> const & right);
	operator()(cute::c2AABB const & aabb)
	{
		std::visit(vis::CollisionAabbManifoldVisitor(m_manifold, aabb), m_right);
	}
	operator()(cute::c2Circle const & circle)
	{
		std::visit(vis::CollisionCircleManifoldVisitor(m_manifold, circle), m_right);
	}
	cute::c2Manifold & m_manifold;
	std::variant<cute::c2AABB, cute::c2Circle> const & m_right;
};
```
*For simplicity I will write the AABB visitor since all that changes if the `cute::c2AABBToAABB()`
```cpp
class CollisionAabbManifoldVisitor
{
	CollisionAabbManifoldVisitor(cute::c2Manifold & manifold, cute::c2AABB const & left);
	operator()(cute::c2AABB const & aabb)
	{
		cute::c2AABBToAABBManifold(m_left, aabb, &m_manifold); // double check function spelling is correct
	}
	operator()(cute::c2Circle const & circle)
	{
		cute::c2AABBToCircleManifold(m_left, circle, &m_manifold); // double check function spelling is correct
	}
	cute::c2Manifold & m_manifold;
	cute::c2AABB const & m_left;
}